### PR TITLE
add ics23 proof support

### PR DESCRIPTION
**Pull request by yutianwu**
This pr will add ics23 proof support for iavl.

### CODE REVIEW

**Review by brilliant-lx — DISMISSED** [review]

File: go.mod [hunk 1]
```diff
@@ -3,25 +3,32 @@ module github.com/tendermint/iavl
 go 1.12
 
 require (
-	github.com/davecgh/go-spew v1.1.1 // indirect
+	github.com/confio/ics23/go v0.7.0
```
Comment (guagualvcha) [inline go.mod:6]: We should not depend `confio/ics23` anymore even though we use our own repo to relace it, it is not a secure lib, and cosmos are not going to use it as well.

File: go.mod [hunk 2]
```diff
@@ -3,25 +3,32 @@ module github.com/tendermint/iavl
...
+	golang.org/x/net v0.0.0-20220812174116-3211cb980234 // indirect
+	golang.org/x/sys v0.1.0 // indirect
+	google.golang.org/protobuf v1.28.1 // indirect
+	gopkg.in/check.v1 v1.0.0-20200227125254-8fa46927fb4f // indirect
+	gopkg.in/yaml.v2 v2.4.0 // indirect
 )
 
 replace github.com/tendermint/go-amino => github.com/binance-chain/bnc-go-amino v0.14.1-binance.2
+
+replace github.com/confio/ics23/go => github.com/bnb-chain/ics23 v0.0.0-20221019083415-6917387c7907
```
Comment (guagualvcha) [inline go.mod:34]: I think we can use `github.com/bnb-chain/ics23` directly.

File: immutable_tree.go [hunk 3]
```diff
@@ -155,6 +155,19 @@ func (t *ImmutableTree) Get(key []byte) (index int64, value []byte) {
 	return root.get(t, key)
 }
 
+// GetWithIndex returns the index and value of the specified key if it exists, or nil and the next index
+// otherwise. The returned value must not be modified, since it may point to data stored within
+// IAVL.
+//
+// The index is the index in the list of leaf nodes sorted lexicographically by key. The leftmost leaf has index 0.
+// It's neighbor has index 1 and so on.
+func (t *ImmutableTree) GetWithIndex(key []byte) (int64, []byte) {
```
Comment (guagualvcha) [inline immutable_tree.go:164]: `func (t *ImmutableTree) GetWithIndex(key []byte) (index int64, value []byte) ` can be more straightforwad.

File: proof_ics23.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+	proof := &ics23.CommitmentProof{
+		Proof: &ics23.CommitmentProof_Nonexist{
+			Nonexist: nonexist,
+		},
+	}
+	return proof, nil
+}
+
+func createExistenceProof(tree *ImmutableTree, key []byte) (*ics23.ExistenceProof, error) {
```
Comment (guagualvcha) [inline proof_ics23.go:78]: The cosmos make `createExistenceProof` an interface of `ImmutableTree`, any reason we change it?

**Review by guagualvcha — APPROVED** [review]

**Pull request merged**